File: carehub-backend/src/main/java/vn/vietduc/carehubbackend/questiongeneration/security/EvaluationPermissions.java
package vn.vietduc.carehubbackend.questiongeneration.security;

import java.util.Set;

public final class EvaluationPermissions {
    public static final String QUESTION_AUTHOR = "QUESTION_AUTHOR";
    public static final String QUESTION_REVIEWER = "QUESTION_REVIEWER";
    public static final String QUESTION_SET_MANAGER = "QUESTION_SET_MANAGER";
    public static final String EXAM_CONFIG_MANAGER = "EXAM_CONFIG_MANAGER";
    public static final String EXAM_PUBLISHER = "EXAM_PUBLISHER";
    public static final String ASSIGNMENT_MANAGER = "ASSIGNMENT_MANAGER";
    public static final String RESULT_VIEWER = "RESULT_VIEWER";
    public static final String AUDIT_VIEWER = "AUDIT_VIEWER";

    public static final Set<String> ALL = Set.of(
            QUESTION_AUTHOR,
            QUESTION_REVIEWER,
            QUESTION_SET_MANAGER,
            EXAM_CONFIG_MANAGER,
            EXAM_PUBLISHER,
            ASSIGNMENT_MANAGER,
            RESULT_VIEWER,
            AUDIT_VIEWER
    );

    private EvaluationPermissions() {
    }
}

File: carehub-frontend/src/features/evaluation/utils/evaluationPermissions.js
import { tokenStorage } from '../../../shared/auth/tokenStorage.js'
import { getPermissionsFromAccessToken, getRolesFromAccessToken } from '../../../shared/auth/jwt.js'

export const EVALUATION_PERMISSION = {
  questionAuthor: 'QUESTION_AUTHOR',
  questionReviewer: 'QUESTION_REVIEWER',
  examConfigManager: 'EXAM_CONFIG_MANAGER',
  examPublisher: 'EXAM_PUBLISHER',
  assignmentManager: 'ASSIGNMENT_MANAGER',
  resultViewer: 'RESULT_VIEWER',
  auditViewer: 'AUDIT_VIEWER',
}

const ALL_EVALUATION_PERMISSIONS = Object.values(EVALUATION_PERMISSION)

export function getCurrentEvaluationAccess() {
  const token = tokenStorage.getAccessToken()
  const roles = getRolesFromAccessToken(token).map(normalize)
  const permissions = getPermissionsFromAccessToken(token).map(normalize)
  const isAdmin = roles.includes('ADMIN')

  return {
    roles,
    permissions,
    isAdmin,
    canAccessEvaluation: isAdmin || ALL_EVALUATION_PERMISSIONS.some((permission) => permissions.includes(permission)),
    hasAny(requiredPermissions = []) {
      if (isAdmin) return true
      if (!requiredPermissions.length) return true
      return requiredPermissions.map(normalize).some((permission) => permissions.includes(permission))
    },
  }
}

function normalize(value) {
  return String(value || '').replace(/^ROLE_/, '').toUpperCase()
}
